k8s/src/main/scala/io/buoyant/k8s/IngressCache.scala

```diff
@@ -0,0 +1,188 @@
+        case Activity.Ok(items) => items
+        case _ => Map.empty[NamespacedName, VarUp[IngressSpec]]
+      }
+      state() = Activity.Ok(items + (name -> Var(item)))
+    }
+
+  def modify(obj: v1beta1.Ingress): Unit =
+    for (name <- getName(obj)) synchronized {
+      log.debug("k8s ns %s modified: %s", namespace.getOrElse(""), name)
+      state.sample() match {
+        case Activity.Ok(snap) =>
+          snap.get(name) match {
+            case None =>
+              log.warning("k8s ns %s received modified watch for unknown resource %s", namespace.getOrElse(""), name)
+            case Some(item) =>
+              mkItem(obj) match {
+                case Some(i) => item() = i
+                case None => state() = Activity.Ok(snap - name)
+              }
+          }
+        case _ =>
+      }
+    }
+
+  def delete(obj: v1beta1.Ingress): Unit =
+    for (name <- getName(obj)) synchronized {
+      log.debug("k8s ns %s deleted: %s", namespace.getOrElse(""), name)
+      state.sample() match {
+        case Activity.Ok(snap) =>
+          for (svc <- snap.get(name)) {
+            state() = Activity.Ok(snap - name)
+          }
+
+        case _ =>
+      }
+    }
+
+  def update(watch: v1beta1.IngressWatch): Unit = watch match {
+    case v1beta1.IngressError(e) => log.error("k8s watch error: %s", e)
+    case v1beta1.IngressAdded(ingress) => add(ingress)
+    case v1beta1.IngressModified(ingress) => modify(ingress)
+    case v1beta1.IngressDeleted(ingress) => delete(ingress)
+  }
+
+  /**
+   * Ingress-specific methods
+   */
+
+  def getName(ingress: v1beta1.Ingress): Option[NamespacedName] =
+    ingress.metadata.flatMap { i =>
+      (i.namespace, i.name) match {
+        case (Some(ns), Some(n)) => Some(NamespacedName(i.namespace, i.name))
+        case _ => None
+      }
+    }
+
+  def mkItem(ingress: v1beta1.Ingress): Option[IngressSpec] = {
+    val namespace = ingress.metadata.flatMap(meta => meta.namespace).getOrElse("default")
+    ingress.spec match {
+      case Some(spec) =>
+        val paths = for (
+          spec <- ingress.spec.toSeq;
+          rules <- spec.rules.toSeq;
+          rule <- rules;
+          http <- rule.http.toSeq;
+          path <- http.paths
+        ) yield {
+          IngressPath(rule.host, path.path.map(Path.read(_)), namespace, path.backend.serviceName, path.backend.servicePort)
+        }
+
+        val fallback = spec.backend.map(b => IngressPath(None, None, namespace, b.serviceName, b.servicePort))
+        Some(IngressSpec(ingress.metadata.flatMap(meta => meta.name), fallback, paths))
+      case None => None
+    }
+  }
+
+  def isPrefix(pfx: Path, str: Path): Boolean = pfx.showElems == str.take(pfx.size).showElems
+  def matchesNs(ns: String): Boolean = namespace.map(_ == ns).getOrElse(true)
+
+  def getMatchingPath(hostHeader: Option[String], requestPath: String): Future[Option[IngressPath]] = {
+    val pathToMatch = Path.read(requestPath)
```

Comment:
Oh! Here I was thinking how nice I am to convert all the uris to fpaths. Seemed like a cleaner way to compare prefixes, but I'm also cool w/ string manip